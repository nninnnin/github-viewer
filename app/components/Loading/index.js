import React, { useEffect } from "react";
import PropTypes from "prop-types";

const styles = {
  content: {
    fontSize: "35px",
    position: "absolute",
    left: "0",
    right: "0",
    marginTop: "20px",
    textAlign: "center",
  },
};

/*

  TODO: 아래 Loading 컴포넌트를 함수형 컴포넌트로 수정하고, `/spec/Loading.spec.js`에 테스트 내용을 보강하세요.

 */
export default class Loading extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      content: props.text,
    };
  }

  componentDidMount() { // 컴포넌트가 마운트되었을 때
    const { speed, text } = this.props;

    this.interval = window.setInterval(() => {
      this.state.content === text + "..."
        ? this.setState({ content: text })
        : this.setState(({ content }) => ({ content: content + "." }));
    }, speed);
  }

  componentWillUnmount() {
    window.clearInterval(this.interval);
  }

  render() {
    return <p style={styles.content}>{this.state.content}</p>;
  }
}

function Loading2 ({ speed, text }) {
  const [content, setContent] = useState('');

  useEffect(() => {


    return function () {
      window.clearInterval(this.interval);
    }
  }, []); // 원하는 것은 마운트 될 때에만 한번 실행

  return (
    <p style={styles.content}>{content}</p>
  );
}

Loading.propTypes = {
  text: PropTypes.string.isRequired,
  speed: PropTypes.number.isRequired,
};

Loading.defaultProps = {
  text: "Loading",
  speed: 300,
};
