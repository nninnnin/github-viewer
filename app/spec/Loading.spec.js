import React from "react";
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';
import { shallow, mount } from "enzyme";
import Loading from "../components/Loading";

beforeEach(() => {
  jest.useFakeTimers();
})

afterEach(() => {
  act(() => {
    jest.runOnlyPendingTimers();
  });
  jest.useRealTimers();
})

test("Snapshot test", () => {
  const TEXT = "hello world";
  const loading = mount(<Loading text={TEXT} />);

  expect(loading).toMatchSnapshot();
});

test("Should have prop name of text and delivered value", () => {
  const TEXT = "PROPTEST";
  const wrapper = mount(<Loading text={TEXT} />);

  expect(wrapper.props().text).toEqual('PROPTEST');
});

test("Loading component should show value of text prop", () => {
  const TEXT = "hello world";
  const loading = shallow(<Loading text={TEXT} />);

  expect(loading.text()).toEqual(TEXT);
});

test('Tests for setInterval callback using advanceTimer', () => {
  const initialText = 'initialText';
  const wrapper = mount(<Loading speed = {500} text = {initialText}/>);

  expect(wrapper.text()).toEqual(initialText);

  act(() => {
    jest.advanceTimersByTime(1600);
  });

  expect(wrapper.text()).toEqual(initialText + '...');
});

test('Tests for pending timers', () => {
  const initialText = 'initialText';
  const wrapper = mount(<Loading speed = {500} text = {initialText}/>);

  expect(wrapper.text()).toEqual(initialText);

  act(() => {
    jest.runOnlyPendingTimers();
  });
  expect(wrapper.text()).toEqual(initialText + '.');
});

test('Tests for pending timers 2', () => {
  const initialText = 'initialText';
  const wrapper = mount(<Loading speed = {500} text = {initialText}/>);

  expect(wrapper.text()).toEqual(initialText);

  act(() => {
    jest.runOnlyPendingTimers();
    jest.runOnlyPendingTimers();
    jest.runOnlyPendingTimers();
  });
  expect(wrapper.text()).toEqual(initialText + '..');
});

test('Tests for pending timers 3', () => {
  const initialText = 'initialText';
  const wrapper = mount(<Loading speed = {500} text = {initialText}/>);

  expect(wrapper.text()).toEqual(initialText);

  act(() => {
    jest.runOnlyPendingTimers();
    jest.runOnlyPendingTimers();
    jest.runOnlyPendingTimers();
    jest.runOnlyPendingTimers();
    jest.runOnlyPendingTimers();
  });
  expect(wrapper.text()).toEqual(initialText + '...');
});
