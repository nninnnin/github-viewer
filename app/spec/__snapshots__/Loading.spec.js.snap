// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading component should show text prop 1`] = `
<p
  style={
    Object {
      "fontSize": "35px",
      "left": "0",
      "marginTop": "20px",
      "position": "absolute",
      "right": "0",
      "textAlign": "center",
    }
  }
>
  Shallow
</p>
`;

exports[`Snapshot test 1`] = `
<Loading2
  speed={300}
  text="hello world"
>
  <p
    style={
      Object {
        "fontSize": "35px",
        "left": "0",
        "marginTop": "20px",
        "position": "absolute",
        "right": "0",
        "textAlign": "center",
      }
    }
  >
    hello world
  </p>
</Loading2>
`;
